src/LeaseManager.cc

```diff
@@ -0,0 +1,307 @@
+ */
+LeaseManager::LeasePreallocator::LeasePreallocator(Context* context,
+                                                  LeaseManager* leaseManager)
+    : WorkerTimer(context->dispatch)
+    , leaseManager(leaseManager)
+{}
+
+/**
+ * The handler performs the lease preallocation.
+ */
+void
+LeaseManager::LeasePreallocator::handleTimerEvent()
+{
+    uint64_t preallocationCount = 0;
+    while (true) {
+        LeaseManager::Lock lock(leaseManager->mutex);
+        preallocationCount = leaseManager->maxAllocatedLeaseId -
+                             leaseManager->lastIssuedLeaseId;
+        if (preallocationCount >= PREALLOCATION_LIMIT)
+            break;
+        leaseManager->allocateNextLease(lock);
+    }
+}
+
+
+/**
+ * Constructor for the LeaseCleaner.
+ * \param context
+ *      Overall information about the RAMCloud server and provides access to
+ *      the dispatcher.
+ * \param leaseManager
+ *      Provides access to the leaseManager to perform cleaning.
+ */
+LeaseManager::LeaseCleaner::LeaseCleaner(Context* context,
+                                         LeaseManager* leaseManager)
+    : WorkerTimer(context->dispatch)
+    , leaseManager(leaseManager)
+{}
+
+/**
+ * This handler performs a cleaning pass on leaseManager.
+ */
+void
+LeaseManager::LeaseCleaner::handleTimerEvent()
+{
+    bool stillCleaning = true;
+    while (stillCleaning) {
+        stillCleaning = leaseManager->cleanNextLease();
+    }
+    // Run once per lease term as some will likely have expired by then.
+    this->start(Cycles::rdtsc() + Cycles::fromNanoseconds(
+            LeaseCommon::LEASE_TERM_US * 1000));
+}
+
+/**
+ * Persist the next available leaseId to external storage=.
+ *
+ * \param lock
+ *      Ensures that caller has acquired mutex; not actually used here.
+ */
+void
+LeaseManager::allocateNextLease(Lock &lock)
+{
+    uint64_t nextLeaseId = maxAllocatedLeaseId + 1;
+    std::string leaseObjName = STORAGE_PREFIX + "/"
+            + format("%lu", nextLeaseId);
+    std::string str;    // TODO(cstlee): Is there a better way set an empty obj?
+    context->externalStorage->set(ExternalStorage::Hint::CREATE,
+                                  leaseObjName.c_str(),
+                                  str.c_str(), 0);
+    maxAllocatedLeaseId = nextLeaseId;
+}
+
+/**
+ * Expire the next lease whose term has elapsed. If the term of the next lease
+ * to be expired has not yet elapsed, this call has no effect.
+ *
+ * \return
+ *      Return true if a lease was able to be cleaned.  Returning false implies
+ *      there are no more leases that can be cleaned at this moment.
+ */
+bool
+LeaseManager::cleanNextLease()
+{
+    Lock lock(mutex);
+    ExpirationOrderSet::iterator it = expirationOrder.begin();
+    if (it != expirationOrder.end() && it->leaseTerm < clock.getTime()) {
+        uint64_t leaseId = it->leaseId;
+        std::string leaseObjName = STORAGE_PREFIX + "/"
```

Comment:
Make a method for computing object names so that you don't have to duplicate this code.